src/main/Scheduler.java

```diff
@@ -0,0 +1,151 @@
+package main;
+import java.time.LocalTime;
+import java.util.ArrayList;
+
+import events.ArrivalEvent;
+import events.FloorEvent;
+import states.ActiveState;
+import states.IdleState;
+import states.SchedulerState;
+
+public class Scheduler implements Runnable {
+	private ArrayList<FloorEvent> floorEvents;
+	private ArrayList<ArrivalEvent> arrivalEvents;
+	private ArrayList<DestinationEvent> destinationEvents; 
+	private MiddleMan middleManFloor;
+	private MiddleMan middleManElevator;
+	private SchedulerState currentState; 
+	
+	
+	/**
+	 * Public constructor to create Scheduler object and instantiate instance variables
+	 * @param middleMan Object to hold and pass events to/from the floor
+	 * @param middleMan2 Object to hold and pass events to/from the elevator
+	 */
+	public Scheduler(MiddleMan middleManFloor, MiddleMan middleManElevator) {
+		this.floorEvents = new ArrayList<FloorEvent>();
+		this.arrivalEvents = new ArrayList<ArrivalEvent>();
+		this.destinationEvents = new ArrayList<DestinationEvent>(); 
+		this.middleManFloor = middleManFloor;
+		this.middleManElevator = middleManElevator;
+		this.currentState = new IdleState(this); 
+	}
+	
+	/**
+	 * Used to communicate with Floor Subsystem and Elevator to receive and send events.
+	 */
+	@Override
+	public void run() {
+		while(true) {
+			
+			ArrivalEvent arrivalEvent; 
+			DestinationEvent destinationEvent; 
+			SchedulerEvent schedulerEvent; 
+			FloorEvent floorEvent = getFloorEvent(), currentFloorEvent;
+			boolean floorEventFlag = false, destinationEventFlag = false, keepGoing = false; 
+			
+			currentState = currentState.handleFloorEvent(floorEvent);
+			currentState = currentState.handleArrivalEvent();
+			currentState = currentState.handleDestinationEvent(destinationEvent);
+			
+			if(currentState.getClass() == ActiveState.class) {
+				if (!floorEvents.isEmpty()) {
+					//TO:DO Check if needed. 
+					floorEvent = floorEvents.get(floorEvents.size() - 1);
+					middleManElevator.putFloorEvent(floorEvent);
+				}
+				
+				if (!arrivalEvents.isEmpty()) {
+					arrivalEvent = arrivalEvents.remove(0);
+					
+					for(FloorEvent fEvent : floorEvents) {
+						if(arrivalEvent.getCurrentFloor() == fEvent.getSource()) {
+							currentFloorEvent = fEvent; 
+							floorEventFlag = true; 
+							break; 
+						}
+												
+					}
+					
+					for(DestinationEvent destEvent : destinationEvents) {
+						if(destEvent.getDestination() == arrivalEvent.getCurrentFloor()) {
+							keepGoing = !destinationEvents.isEmpty();
+							//set the floor event to null since thats what elevator wants. 
+							currentFloorEvent = null; 
+							destinationEventFlag = true; 

```

Comment:
I woudln't make currentFloorEvent = null. If you're at a floor that is a source floor AND destination floor, then the scheduler event should have keepGoing = true, isAtDestination=True, isAtSource = true, and floorevent != null (because you need it to service the source floor)

Reply:
fixed

Reply:
I don't see where elevatorKeepsGoing is being set in the code (rn I think its false all the time)

Reply:
Gets set here: 					//checks if the elevator should keep going. 
					elevatorKeepsGoing = !destinationEvents.isEmpty();